src/coreclr/vm/peimagelayout.cpp

```diff
@@ -279,7 +281,14 @@ void PEImageLayout::ApplyBaseRelocations(bool relocationMustWriteCopy)
             switch (fixup>>12)
             {
             case IMAGE_REL_BASED_PTR:
+#if defined(__APPLE__) && defined(HOST_ARM64)
+                {
+                    ExecutableWriterHolder<TADDR> addressWriterHolder((TADDR *) address, sizeof(TADDR));
```

Comment:
It looks very expensive performance-wise to take the writer holder for every reloc.

Reply:
Please note that this is only for arm64 macOS, so this is cheap as it only calls the API to switch the current CPU mode. However, it would probably be better to use the holder around the whole function or call the PAL_JitWriteProtect at the beginning and end of this function to be more explicit with the fact that this is apple specific.

Reply:
You're right — I did a quick calculation and it's about 9 times slower in average on applying relocations on the images used by the basic console template. Though we are talking really small numbers: 140us vs 940us for all the relocations on a M1 Max. I have a version where I take the writer holder once per section and it goes down to 100us on the same machine. The code is slightly more complex and harder to follow, but I'm happy to commit it to this branch if you'd like.

Reply:
I have reimplemented it with @janvorli's suggestion of using `PAL_JitWriteProtect` at the beginning and end of `PEImageLayout::ApplyBaseRelocations`. It's as simple and readable as my original change, and much faster.
Thanks to both Jans!